spinn_front_end_common/interface/interface_functions/spalloc_machine_generator.py

```diff
@@ -0,0 +1,148 @@
+    """
+
+    __slots__ = []
+
+    MAX_RESTART_SIZE_IN_BOARDS = 24
+
+    def __call__(
+            self, spalloc_server, spalloc_user, downed_chips, downed_cores,
+            downed_links, sick_boards, n_chips=None, n_boards=None,
+            spalloc_port=None, spalloc_machine=None, max_sdram_size=None,
+            repair_machine=False, ignore_bad_ethernets=True,
+            default_report_directory=None):
+        """
+        :param str spalloc_server:
+        :param str spalloc_user:
+        :param set(tuple(int,int)) downed_chips:
+        :param set(tuple(int,int,int)) downed_cores:
+        :param set(tuple(int,int,int)) downed_links:
+        :param iter(str) sick_boards:
+        :param int n_chips:
+        :param int n_boards:
+        :param int spalloc_port:
+        :param str spalloc_machine:
+        :rtype: tuple(str, int, None, bool, bool, None, None,
+            MachineAllocationController)
+        """
+
+        machine_details = None
+        previous_controller = None
+
+        spalloc_allocator = SpallocAllocator()
+        machine_generator = MachineGenerator()
+
+        while not machine_details:
+            results = spalloc_allocator(
+                spalloc_server, spalloc_user, n_chips, n_boards, spalloc_port,
+                spalloc_machine)
+            hostname, board_version, bmp_details, reset_machine_on_start_up, \
+                auto_detect_bmp, scamp_connection_data, boot_port_num, \
+                machine_allocation_controller = results
+
+            # Now we have a new job release the older one
+            if previous_controller:
+                previous_controller.close()
+
+            try:
+                machine_details, txrx = machine_generator(
+                    hostname, bmp_details, downed_chips, downed_cores,
+                    downed_links, board_version, auto_detect_bmp,
+                    scamp_connection_data, boot_port_num,
+                    reset_machine_on_start_up, max_sdram_size, repair_machine,
+                    ignore_bad_ethernets, default_report_directory)
+            except SpinnMachineCorruptionException as ex:
+                if n_boards and n_boards > self.MAX_RESTART_SIZE_IN_BOARDS:
+                    raise
+                if n_chips and n_chips > self.MAX_RESTART_SIZE_IN_BOARDS * 48:
+                    raise
+                machine_generator = MachineGenerator()
+                machine_details, txrx = machine_generator(
+                    hostname, bmp_details, downed_chips, downed_cores,
+                    downed_links, board_version, auto_detect_bmp,
+                    scamp_connection_data, boot_port_num,
+                    reset_machine_on_start_up,
+                    max_sdram_size, repair_machine, ignore_bad_ethernets,
+                    default_report_directory)
+                known = False
+                for sick_ip in sick_boards:
+                    if sick_ip in ex.ipaddress:
```

Comment:
This logic seems backwards. We want to exclude known bad boards from reporting, not work through the list of bad boards and check if any of them are in there. Though I suppose recovering if any of them were known is somewhat sensible too.

Need to be clear exactly what conditions trigger reporting and what trigger retrying; they might not be the same!